kube/src/api/admission.rs

```diff
@@ -0,0 +1,223 @@
+//! Contains types for implementing admission controllers.
+//!
+//! For more information on admission controllers, see:
+//! https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/
+//! https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/
+
+use crate::{
+    api::{DynamicObject, GroupVersionKind, GroupVersionResource, Resource, TypeMeta},
+    Error,
+};
+
+use std::{collections::HashMap, convert::TryInto};
+
+use k8s_openapi::{
+    api::authentication::v1::UserInfo,
+    apimachinery::pkg::{apis::meta::v1::Status, runtime::RawExtension},
+};
+use serde::{Deserialize, Serialize};
+
+/// The top level struct used for Serializing and Deserializing AdmissionReview
+/// requests and responses.
+#[derive(Serialize, Deserialize, Clone, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct AdmissionReview<T: Resource> {
+    /// Contains the API version and type of the request.
+    #[serde(flatten, default)]
+    pub typ: TypeMeta,
+    /// Describes the attributes for the admission request.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub request: Option<AdmissionRequest<T>>,
+    /// Describes the attributes for the admission response.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub response: Option<AdmissionResponse>,
+}
+
+impl<T: Resource> TryInto<AdmissionRequest<T>> for AdmissionReview<T> {
+    type Error = Error;
+
+    fn try_into(self) -> Result<AdmissionRequest<T>, Self::Error> {
+        match self.request {
+            Some(req) => Ok(req),
+            None => Err(Error::RequestValidation(
+                "invalid AdmissionRequest. expected Some but got None".to_owned(),
+            )),
+        }
+    }
+}
+
+/// An incoming [`AdmissionReview`] request.
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct AdmissionRequest<T: Resource> {
+    /// An identifier for the individual request/response. It allows us to
+    /// distinguish instances of requests which are otherwise identical (parallel
+    /// requests, requests when earlier requests did not modify, etc). The UID is
+    /// meant to track the round trip (request/response) between the KAS and the
+    /// webhook, not the user request. It is suitable for correlating log entries
+    /// between the webhook and apiserver, for either auditing or debugging.
+    pub uid: String,
+    /// The fully-qualified type of object being submitted (for example, v1.Pod
+    /// or autoscaling.v1.Scale).
+    pub kind: GroupVersionKind,
+    /// The fully-qualified resource being requested (for example, v1.pods).
+    pub resource: GroupVersionResource,
+    /// The subresource being requested, if any (for example, "status" or
+    /// "scale").
+    pub sub_resource: Option<String>,
```

Comment:
I would have thought a lot of these required `#[serde(default)]`, are they explicity passed as null in the json?

Reply:
I swapped between using `#[serde(default)]` and `Option<>`s a couple times, but finally just ended up keeping the options as that seemed to be what all the `k8s_openapi` types did for the optional types. The Go structs have non-pointers for everything but nested structs, but nil struct pointers will never be serialized (see `,omitempty` here: https://github.com/kubernetes/api/blob/master/admission/v1/types.go#L65). So, there should only ever be missing fields, not null.

Usability wise, I definitely prefer the `#[serde(default)]`. So I'd be happy to swap back to that.

Reply:
Unless you meant `Option`s should fail when there's no `#[serde(default)]` and the field isnt provided

Reply:
Yeah, it was originally the latter point I was afraid of. Since they are Options which can be omitted, i think we should probably put `#[serde(default)]` on them.

As for the first point; the `omitempty` on listable types basicaly means that kubernetes does not distinguish between _empty_ and _missing_, so `#[serde(default)` makes sense on listable container types in those cases. (There's an open issue on k8s-openapi for that). You don't have a lot of them, only see `audit_annotations`.

Reply:
just to confirm, you want `#[serde(default)]` on all of the option types?

Reply:
Yeah, at least in the Request please. I'll comment on the special case explicitly.

Reply:
unfortunately doesn't work on the fields containing the `<T: Resource>` even though it's wrapped in an option. It still requires the trait to inhert `Default` :confused:

Reply:
Ah, but that's fine. That's the `object` or `old_object`, and we don't really want those defaulted. Looks all good to me now!